File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/replace/ReplaceText.java
package com.kite9.k9server.command.xml.replace;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;

public class ReplaceText extends AbstractReplaceCommand<Element, String> {
	
	public enum PreserveChildElements { BEFORE, AFTER, NONE };
	
	public PreserveChildElements preserve = PreserveChildElements.AFTER;
	
	private List<Element> collectChildren(Element e) {
		if (preserve == PreserveChildElements.NONE) {
			return Collections.emptyList();
		} else {
			List<Element> out = new ArrayList<Element>();
			NodeList nl = e.getChildNodes();
			for (int i = 0; i < nl.getLength(); i++) {
				if (nl.item(i) instanceof Element) {
					out.add((Element) nl.item(i)); 
				}
			}
			return out;
		}
	}

	@Override
	protected String getFromContent(ADLDom context) {
		return from;
	}

	@Override
	protected String getToContent(ADLDom context) {
		return to;
	}

	@Override
	protected Element getExistingContent(ADLDom in) {
		Element on = findFragmentElement(in.getDocument(), fragmentId);
		return on;
	}

	@Override
	protected void doReplace(ADLDom adl, Element e, String toContent, String fromContent) {
		List<Element> childElements = collectChildren(e);
		
		e.setTextContent(toContent);
		
		if (preserve == PreserveChildElements.BEFORE) {
			Collections.reverse(childElements);
			childElements.stream().forEach(c -> e.insertBefore(c, null));
		} else if (preserve == PreserveChildElements.AFTER) {
			childElements.stream().forEach(c -> e.appendChild(c));
		}
		
		LOG.info("Processed replace text of "+fragmentId);
	}

	@Override
	protected Mismatch same(Element existing, String with) {
		String eText = existing.getTextContent();
		String eTextReplaced = eText.replaceAll("\\s", "");
		String withReplaced = with.replaceAll("\\s", "");
		switch (preserve) {
		default:
		case AFTER:
			return check(eTextReplaced.startsWith(withReplaced), eText, with);
		case BEFORE:
			return check(eTextReplaced.endsWith(withReplaced), eText, with);
		case NONE:
			return check(eTextReplaced.equals(withReplaced), eText, with);
		}
	}

	private Mismatch check(boolean ok, String eText, String with) {
		return ok ? null : () -> "Text not same: '"+eText+"' and '"+with+"'";
	}
}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/pipeline/command/xml/replace/ReplaceText.kt
package com.kite9.pipeline.command.xml.replace

import com.kite9.pipeline.adl.holder.pipeline.ADLDom
import com.kite9.pipeline.command.Command.Mismatch
import com.kite9.pipeline.command.CommandContext
import org.w3c.dom.Element
import java.util.*

class ReplaceText : AbstractReplaceCommand<Element, String>() {

    enum class PreserveChildElements {
        BEFORE, AFTER, NONE
    }

	var preserve = PreserveChildElements.AFTER

    private fun collectChildren(e: Element): List<Element?> {
        return if (preserve == PreserveChildElements.NONE) {
            emptyList<Element>()
        } else {
            val out: MutableList<Element?> = ArrayList()
            val nl = e.childNodes
            for (i in 0 until nl.length) {
                if (nl.item(i) is Element) {
                    out.add(nl.item(i) as Element)
                }
            }
            out
        }
    }

    override fun getFromContent(context: ADLDom, ctx: CommandContext): String {
        return from!!
    }

    override fun getToContent(context: ADLDom, ctx: CommandContext): String {
        return to!!
    }

    override fun getExistingContent(o: ADLDom, ctx: CommandContext): Element {
        return findFragmentElement(o.document, fragmentId, ctx)
    }

    override fun doReplace(on: ADLDom, site: Element, toContent: String, fromContent: String, ctx: CommandContext) {
        val childElements = collectChildren(site)
        site.textContent = toContent
        if (preserve == PreserveChildElements.BEFORE) {
            Collections.reverse(childElements)
            childElements.stream().forEach { c: Element? -> site.insertBefore(c, null) }
        } else if (preserve == PreserveChildElements.AFTER) {
            childElements.stream().forEach { c: Element? -> site.appendChild(c) }
        }
        ctx.log("Processed replace text of $fragmentId")
    }

    protected override fun same(existing: Element, with: String): Mismatch? {
        val eText = existing.textContent
        val eTextReplaced = eText.replace("\\s".toRegex(), "")
        val withReplaced = with.replace("\\s".toRegex(), "")
        return when (preserve) {
            PreserveChildElements.AFTER -> check(eTextReplaced.startsWith(withReplaced), eText, with)
            PreserveChildElements.BEFORE -> check(eTextReplaced.endsWith(withReplaced), eText, with)
            PreserveChildElements.NONE -> check(eTextReplaced == withReplaced, eText, with)
            else -> check(eTextReplaced.startsWith(withReplaced), eText, with)
        }
    }

    private fun check(ok: Boolean, eText: String, with: String): Mismatch? {
        return if (ok) null else Mismatch { "Text not same: '$eText' and '$with'" }
    }
}